### ISSUE TITLE
session withTransaction with Promise.all won't revok all the actions if Promise.all throw an error

### ISSUE THREAD

**Issue by dumbokitty** [Automattic/mongoose#8713]
**Do you want to request a *feature* or report a *bug*?**
bug
**What is the current behavior?**

```
   //User schema is this 
    const userSchema = new Schema({ 
       name: {
         type: String,
          required: true,
       },
       gender: {
          type: String,
          required: true
       }
    })

    const userData = [{name: 'user1' , gender: 'f'},  { name: 'user2'}];  //  an array with user data

    const session = await User.startSession();

    await session.withTransaction(async()= {

    await Promise.all(userData.map(async (oneUser)=> {

      const userRecord = new User(oneUser); //Some irrelavant code is skipped

       await userRecord,save({session}}))

    })
```
 
**If the current behavior is a bug, please provide the steps to reproduce.**

**What is the expected behavior?**
   Since userData[1] misses the "gender " field, it would fail the model validation and wound't get saved properly, I would expect withTransaction will be able to revoke the saving of userData[0] too. However, I found that userDatap[0]  did get to saved into the db.  Moreover,  if I don't use session.withTransaction, following code seems to give me the expected behavior that none of the userData get saved 
    

```
    const session = await User.startSession();
     await session.withTransaction();
     try {
         await Promise.all(userData.map(async (oneUser)=> {

      const userRecord = new User(oneUser); //Some irrelavant code is skipped

       await userRecord,save({session}}));
       await session.commitTransaction();
       session.endSession();
     } catch(error) {
        await session.abortTransaction();
        session.endSession();
        throw error
     }
```

However, using this will sometimes give me the mismatch id/number error which I've read about in the issues here where the answer is to user withTransaction helper.  I am wondering if withTransaction error handles Promise.all differently?  

**What are the versions of Node.js, Mongoose and MongoDB you are using? Note that "latest" is not a version.**
  Mongoose: 5.7.11
  MongoDB: 4.2.1
   node :10.16.0

**Comment by vkarpov15** [comment]
I'm unable to repro this, the below script demonstrates that no users are saved. Please modify the below script to demonstrate your issue.

```javascript
const mongoose = require('mongoose');

mongoose.set('useFindAndModify', false);

const { Schema } = mongoose;

run().catch(err => console.log(err));

async function run() {
  await mongoose.connect('mongodb://localhost:27017/test?replicaSet=rs', {
    useNewUrlParser: true,
    useUnifiedTopology: true
  });

  await mongoose.connection.dropDatabase();

  const userSchema = new Schema({
    name: {
      type: String,
      required: true,
    },
    gender: {
      type: String,
      required: true
    }
  });
  const User = mongoose.model('User', userSchema);

  const userData = [{name: 'user1' , gender: 'f'},  { name: 'user2'}];  //  an array with user data

  const session = await User.startSession();

  await session.withTransaction(async() => {
    await Promise.all(userData.map(async (oneUser) => {
      const userRecord = new User(oneUser);
      await userRecord.save({session});
    }));
  }).catch(err => console.log('Err', err));

  console.log(await User.find()); // Prints "[]", no users saved
  console.log('done');
} 
```

**Comment by dietergeerts** [comment]
I was able to reproduce this in our application. We were using `Promise.all`, and 1 out of the 3 `save()` statements failed. A transaction session was on all 3 documents, and it was the same session. The data in the database became corrupt, as the one save was persisted, and the other 2 were not. When transforming the `Promise.all` to `Promise.allSettled`, it did work. Though I have to mention that we set the session on the document in a different way than through the query, so maybe in the deep internals of mongoose it matters.
